### PR TITLE
Fix YUV alignment when width is not multiple of 64 pixels

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by lizhong1008** [comment]
@ViGill , thanks for your patch. The branch  intel_ffmpeg-3.4.1 is not stable and currently it is created for development (many patches are in the list to be merged).  
BTW, could you specify how to reproduce the issue this patch to fix?  It is recommended to create a github issue with detail reproduce steps.

**Comment by ghost** [comment]
Hi,
you can try to encode a file that has 720 as width.
For instance: https://ufile.io/znn0z
You need to encode in HW but decode in SW, and use VPP scaling.

**Comment by ghost** [comment]
Hi,

any feedback on this PR ?

**Comment by lizhong1008** [comment]
@ViGill , sorry I can download the clip from https://ufile.io/znn0z.  
Could you please upload it to google drive?   And providing your testing command should be better.

**Comment by ghost** [comment]
Hi,

the file was removed, please use https://ufile.io/b06ur instead (let me know if it fails again, I'll use google drive if it does).
The commande line is for instance:
ffmpeg -i spam.avi -an -vcodec h264_qsv -vf [main]vpp_qsv=w=512:h=288[out] out.mp4

**Comment by lizhong1008** [comment]
@ViGill , still can't open the link https://ufile.io/b06ur.  
But I tried another video: ./ffmpeg -i /samba/anonymous/Videos/big_buck_bunny_480p_surround-fix.avi -an -vcodec h264_qsv -vf [main]vpp_qsv=w=512:h=288[out] out.mp4. It failed just like issue #19 
[Parsed_vpp_qsv_0 @ 0x398b540] No hw context provided.
[Parsed_vpp_qsv_0 @ 0x398b540] Failed to configure output pad on Parsed_vpp_qsv_0
Error reinitializing filters!
Failed to inject frame into filter network: Invalid argument
Error while processing the decoded data for stream #0:0

Adding "-init_hw_device qsv=foo -filter_hw_device foo", it can work well. 
The width of this clip is not 720 but 854 (also not 64 (or 32) bytes aligned).
mediainfo /samba/anonymous/Videos/big_buck_bunny_480p_surround-fix.avi
General
Complete name                            : /samba/anonymous/Videos/big_buck_bunny_480p_surround-fix.avi
Format                                   : AVI
Format/Info                              : Audio Video Interleave
File size                                : 210 MiB
Duration                                 : 9 min 56 s
Overall bit rate                         : 2 958 kb/s

Video
ID                                       : 0
Format                                   : MPEG-4 Visual
Format profile                           : Simple@L1
Format settings, BVOP                    : No
Format settings, QPel                    : No
Format settings, GMC                     : No warppoints
Format settings, Matrix                  : Default (H.263)
Codec ID                                 : FMP4
Duration                                 : 9 min 56 s
Bit rate                                 : 2 500 kb/s
Width                                    : 854 pixels
Height                                   : 480 pixels
Display aspect ratio                     : 16:9
Frame rate                               : 24.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.254
Stream size                              : 178 MiB (85%)
Writing library                          : Lavc51.44.0

**Comment by ghost** [comment]
Hi,

yes, it's because of the missing parameters.
Try:

ffmpeg -init_hw_device qsv=foo -filter_hw_device foo -i spam.avi -an -vcodec h264_qsv -vf [main]vpp_qsv=w=512:h=288[out] out.mp4

with the attached file
[spam.avi.zip](https://github.com/Intel-FFmpeg-Plugin/Intel_FFmpeg_plugins/files/1810678/spam.avi.zip)

**Comment by ghost** [comment]
Hi,

this fix was fine on my version but it looks like it doesn't fix the problem here.
Please let me know...

Thanks

**Comment by ghost** [comment]
Hi,

Ok I found the issue, actually in the code I use, we align on 128 and 64 bits:

```
> diff --git a/libavfilter/qsvvpp.c b/libavfilter/qsvvpp.c
> index 0ba9c0e..296b5a4 100644
> --- a/libavfilter/qsvvpp.c
> +++ b/libavfilter/qsvvpp.c
> @@ -168,7 +168,9 @@ static QSVFrame *submit_frame(FFQSVVPPContext *s, AVFilterLink *inlink, AVFrame
>          qsv_frame->surface = (mfxFrameSurface1*)qsv_frame->frame->data[3];
>      } else {
>          /* make a copy if the input is not padded as libmfx requires */
> -        if (picref->height & 31 || picref->linesize[0] & 31) {
> +        const AVPixFmtDescriptor *desc = av_pix_fmt_desc_get(picref->format);
> +        if (picref->height & 31 || picref->linesize[0] & 31
> +            || (!strcmp(desc->name, "yuv420p") && picref->linesize[0] & 63)) {
>              qsv_frame->frame = ff_get_video_buffer(inlink,
>                                              FFALIGN(inlink->w, 128),
>                                              FFALIGN(inlink->h, 64));
```

I will update the PR

File: libavfilter/qsvvpp.c [hunk 1]
```diff
@@ -300,10 +300,12 @@ static QSVFrame *submit_frame(QSVVPPContext *s, AVFilterLink *inlink, AVFrame *p
         qsv_frame->surface = (mfxFrameSurface1 *)qsv_frame->frame->data[3];
     } else {
         /* make a copy if the input is not padded as libmfx requires */
-        if (picref->height & 31 || picref->linesize[0] & 31) {
+        const AVPixFmtDescriptor *desc = av_pix_fmt_desc_get(picref->format);
+        if (picref->height & 31 || picref->linesize[0] & 31
+            || (!strcmp(desc->name, "yuv420p") && picref->linesize[0] & 63)) {
```
Comment (lizhong1008) [inline libavfilter/qsvvpp.c:305]: Only yuv420p format needs this patch or all format of system memory?  And currently your patch will change the original logic for the case "not yuv420p , 32bytes aligned but not 64 bytes aligned".

File: libavfilter/qsvvpp.c [hunk 2]
```diff
@@ -300,10 +300,12 @@ static QSVFrame *submit_frame(QSVVPPContext *s, AVFilterLink *inlink, AVFrame *p
...
             qsv_frame->frame = ff_get_video_buffer(inlink,
-                                                   FFALIGN(inlink->w, 32),
-                                                   FFALIGN(inlink->h, 32));
+                                                   FFALIGN(inlink->w, 128),
+                                                   FFALIGN(inlink->h, 64));
```
Comment (lizhong1008) [inline libavfilter/qsvvpp.c:308]: For the case of yuv420p:
1. picref->linesize[0] is checked as 64 bytes aligned, but here 128 bytes aligned is required for width. 
2.  picref->height hasn't been checked, but here 64 bytes aligned is required. 
3.  Is the 128x64 alignment requirement written in MSDK documentation?

**Comment by ViGill** [comment]
Hi,

I am not sure about the required alignment, it was in the code I had (I just added the check to enter in this case).

BR

File: libavfilter/qsvvpp.c (hunk 1 above)
Comment (ghost) [reply libavfilter/qsvvpp.c:305]: Hi, 
FFmpeg aligns linesize of all components (Y,U,V) with 32 bits separately. When width = 720, after alignment the Y’s linesize is 736, and U, V’s line size is 384 and “736 != 384 * 2”. But mfxFrameSurface1 in current mSDK has only one linesize field for Y component; and I think it use half of Y’s linesize to be U, V’s linesize, and it’s the root cause of the issue.
When  using both HW Decoder and encoder, the frame buffer is vasurface and use NV12 format, and FFmpeg’s alignment won’t be used. So there is no such issue.

File: libavfilter/qsvvpp.c (hunk 2 above)
Comment (ghost) [reply libavfilter/qsvvpp.c:308]: Hi,

I am not sure about the required alignment, it was in the code I had (I just added the check to enter in this case).

BR

File: libavfilter/qsvvpp.c (hunk 1 above)
Comment (lizhong1008) [reply libavfilter/qsvvpp.c:305]: It makes sense.  but I think it should be change to FFALIGN(inlink->w, 64), and no change for inlink->h.

Comment (ghost) [reply libavfilter/qsvvpp.c:305]: if it works no problem

Comment (ghost) [reply libavfilter/qsvvpp.c:305]: Hi, any feedback ?

**Comment by lizhong1008** [comment]
Modified and pushed. Please let me know if any issue.

**Comment by ghost** [comment]
Thanks I will!

**Pull request closed without merging**